README
                                  README

 
   Welcome to PyEdPro. The motivation for this project was to create a modern
multi-platform editor. Simple, powerful, configurable, extendable.

 This project is a successor of pyedit, after porting it to GTK3.

  Working platforms currently are: Win 7 .. Win 10 Centos 6 .. 7 Ubunto 14 ... 16
(all linux platforms) Msys, Msys2, Mingw, Raspberry PI ... practically
 anywhere PyGTK can run.
 
  I mainly run it in MSYS2, Ubuntu and the Raspberry-Pi, and it behaves consistantly
on all three platforms.
 
   PyEdPro.py has macro recording/play, search/replace, functional navigation,
comment/string spell check, auto backup, persistent undo/redo, auto complete,
auto correct, syntax check, spell suggestion ... and a lot more.
   
   It is fast, it is extendable, as python lends itself to easy tinkering. The
editor has a table driven key mapping. One can easily edit the key map in
keyhand.py, and the key actions in acthand.py

The default key map resembles gedit / wed / etp / brief. Full ASCII; 
 Any fixed font can be configured. 

  See KEYS file for the list of keyboard shortcuts or press F3 in the
editor or look at the file in pyedlib/KEYS.

  On initial start, PyEdPro shows a left pane and a top pane. The left pane 
is for function summary and the top pane is for double view of the same file.
(to see the caller and the callee) These panes can be hidden with the mouse by 
dragging on their handle, or by the key combination Alt-Q (Shift-Alt-Q for
the left pane) 

 PyEdPro remembers a lot about your editing. Loaded files, cursor positions,
fonts, font size, colors, search strings, goto numbers, undo / redo info.
It is all stored in ~/.PyEdPro. You may safely delete that directory to start
 PyEdPro with no memory of what has been done.

  Starting PyEdPro with no command line arguments will put you back to the 
 previous session, exactly as you left off.

 The editor will work on Windows, and can open UNIX and Windows files
transparently. It will save the file as the current platform's native 
convention dictates.

  Developer's note: in order to make PyEdPro multi platform, we save
the configuration info into a SQLite database in the ~/.PyEdPro directory.

 Contributors are welcome. Requires pygtk.

 
  Peter Glen








HISTORY
                              HISTORY

 This is a non exhaustive list of changes in the PyEdPro project. Git has
 a more complete list of logs.

    Mon Sep 2 2019
    Reload yes/no dialog, old buffer saved in backup dir

    Sun Sep 1 2019
    fixed cdir on load

    Mon Aug 12 2019
    added terminal right click shortcut

  may-08-2019   (0.71)

    Tested OK on running it in a windows DOS box.

    o Some path inconsistencies (corrected with os.sep)
    o Killed controlling 'terminal' aka DOS box for cleaner operation
    o Updated About dialog to show build date and platform
    o Pop-up for r-click spell check on windows corrected (went out of scope)

  mar-29-2019

  .h files parsed as .c
  multi select on custom file load (alt-o)
  fixed coloring anomaly on color on / off

  mar-05-2019    (0.70)

    Mouse scroll added shift / ctrl to accelerate
    Key Up/Down added shift / ctrl to accelerate
    Still V.67

 feb-13-2019

    Solved jump on parse local
    solved FIFO on search Alt-Left Alt-Right
    V.67

 feb-03-2019

    Added readme section for installing python-gi-cairo

UPDATES:

   PyGTK 3 changed a lot of things ... workig on port

   PyGTK 2.24 (latest at the time of writing) has a bug on not delivering
double click to the window. Down-graded to 2.22.x, works flawlessly.

   Lots of functions added. Attempted to make it tablet friendly. Drag
on screen now does what you would expect from a tablet.

   My laptop keyboard broke, so I added Ctrl-H for left arrow j, k, l
for the other arrows (like in vim) so the defective keys are substituted
with vi compatible navigation.

  On request of one of my clients, I needed to create a time sheet. So I
extended PyEdPro to create a work log.

The log contains the following elements:

 Start time, End time, Time of file Open, time of file Save.
The files have been created to be machine parsable.

 Example log entry:

Action: Started PyEdPro
On: Tue May 08 16:44:37 2018

Action: Ended PyEdPro
On: Tue May 08 17:14:21 2018
Time diff: 1784 0:29

 Example timesheet entry:

 Action: Opened File C:\msys32\home\peterglen\esp\esp-code\wclock\v010\common\v000\httpd.c
On: Thu Jun 28 14:49:13 2018
Delta: 0
Date: 1530211753 httpd.c Opened

Action: Closed File C:\msys32\home\peterglen\esp\esp-code\wclock\v010\common\v000\httpd.c
On: Thu Jun 28 14:49:18 2018
Delta: 5
Date: 1530211759 httpd.c Closed

 The files are located at ~/.PyEdPro/log/account.txt and ~/.PyEdPro/log/timesheet.txt
 (the '~/' stands for your home directory. On windows it is
 C:\Users\<UserName>\ ...) Example: 'C:\Users\PeterGlen\.PyEdPro\'

UPDATES:

  Implemented multiple clipboards. The key combination Ctrl-0 to Ctrl-8 switches
clipboard buffers. Switching to buffer 9 (Ctrl-9) yields the sum of all clipboards
from 1-8. This is useful on collecting parts of the file, and unifying it
with one keystroke.  (Like collecting function headers.) The first clipboard buffer
(Ctrl-0) is the same as the OS clipboard.

UPDATES:

  sep/16/2018   Port to GTP3 started
  oct/10/2018   Port stable, almost done

Pre GIT:

v003:
    Misc
v004:
    Char add, nav, select
v005:
    Added find
v006:
    Added pos persistence, Added pane persistence
    Open, New, Save added
v007:
    Added undo / redo, Tab / sel tab
    Auto complete
v008:
   Cut / Paste
v009:
    Spell check
v010:
    Undo / Pedo
v011:
    Begun to use to edit the editor itself
v012:
    Install etc ...
v013:
    Converted gconf to sqlite
v014:
    Swap words, Cleanup
v015:
    Spell, Syntax check

v020:
    Misc

v021:
    Corrected TAB display
    Search / Replace backwards for correctness
    Auto coloring off if not "c","py"

v022:
    Search / Replace corrected for same line match

v023:
    Auto coloring on if sh
    Eradicated pesky \n from messing up pango

v024:
   Some cleanup

v025:
    Writing to a read only file now has a better error message
    Added a long line wrapper. Wrapping to col 78 or less

v026:
    White background forced. It's an editor ... no fancy coloring needed.
    If you want custom color see variable BGCOLOR

v027:
    Start from current dir on open / save as ... It was inconsistent
    between platforms, current directory forced.

 ....
 ....

v033:
    Lots of new features, cleanup. Tablet like mouse handling. Drag now behaves as
    one would expect it on a tablet.

v034:
    VIM like navigation keys. Tabbing through docs on Alt-arrow(s), Misc fixes

v042:

  Moved to pyedpro, version number restarted at 043.







pyedpro.py
#!/usr/bin/env python3

'''
This is open source text editor. Written on python. The motivation for
this project was to create a modern multi-platform editor. Simple,
powerful, configurable, extendable. To run this module without
installation put the supporting files in the 'pedlib'
subdirectory under the main file's direcory.
'''

from __future__ import absolute_import
from __future__ import print_function

import os
import sys
import getopt
import signal

# ------------------------------------------------------------------------
# This is open source text editor. Written on python. The motivation for
# this project was to create a modern multi-platform editor.
# Simple, powerful, configurable, extendable.
#
# This project is a successor of pyedit.py
#
# Pyedpro functions near identical on Linux / Windows / Mac / Raspberry PI
#
# Pyedpro has:
#
#    o  macro recording/play,
#    o  search/replace,
#    o  functional navigation,
#    o  comment/string spell check,
#    o  full spell check, spell suggestion dialog
#    o  auto backup,
#    o  persistent undo/redo,  (undo beyond last save)
#    o  auto complete, auto correct,
#    o
#    o  ... and a lot more.
#
# It is fast, it is extendable. The editor has a table driven key mapping.
# One can easily edit the key map in keyhand.py, and the key actions
# in acthand.py The default key map resembles gedit / wed / etp / brief

# History:  (recent first, incomplete list)
#
# jul/19/2018       Coloring for spell check, Trigger by scroll, more dominant color
# Jul/xx/2018       Update README, KEYS.TXT
# Jun/xx/2018       Log Files for time accounting.
# Jun/08/2020       Menu control / Headerbar / Version update
# Mon 28.Sep.2020   Reshuffled imports pylint
# Fri 25.Dec.2020   Added web view, m4 filter md2html filterRelese ready

# ASCII text editor, requires pyGtk. (pygobject)
# See pygtk-dependencies for easy install of dependencies.
# See also the INSTALL file.

import gettext
gettext.bindtextdomain('pyedpro', './locale/')
gettext.textdomain('pyedpro')
_ = gettext.gettext
#locale.setlocale(locale.LC_ALL, '')

base = os.path.realpath(__file__)
#print("file", os.path.dirname(base))
os.chdir(os.path.dirname(base))

import pedlib.pedconfig as pedconfig
import pedlib.pedwin as pedwin
import pedlib.pedsql as pedsql
import pedlib.pedlog as pedlog
import pedlib.pedutil as pedutil

import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk

#print("domain", gettext.textdomain)

VERSION = 2.01
BUILDDATE = "Thu 01.Apr.2021"
PROGNAME  = "PyEdPro"

SHOW_TIMING = 0
SHOW_CONFIG = 0
CLEAR_CONFIG = 0
USE_STDOUT = 0

# ------------------------------------------------------------------------

def main(projname, strarr):

    ''' Rotate around this axis '''

    if pedconfig.conf.verbose:
        print(PROGNAME, "running on", "'" + os.name + "'", \
            "GTK", Gtk._version, "PyGtk", \
               "%d.%d.%d" % (Gtk.get_major_version(), \
                    Gtk.get_minor_version(), \
                        Gtk.get_micro_version()))

    signal.signal(signal.SIGTERM, terminate)
    mainwin = pedwin.EdMainWindow(None, None, strarr)
    pedconfig.conf.pedwin = mainwin

    # Create log window
    pedlog.create_logwin()

    if projname:
        mainwin.opensess(projname)

    #if strarr:
    #    print("opening files", strarr)
    #    for aa in strarr:
    #        if os.path.isfile(aa):
    #            mainwin.openfile(aa)

    Gtk.main()

def xversion():
    ''' Offer version number '''
    print("Version", pedconfig.conf.version)
    sys.exit(1)

def xhelp():
    ''' Offer Help '''
    print()
    print(PROGNAME, _("Version: "), pedconfig.conf.version)
    print(_("Usage: ") + PROGNAME + _(" [options] [[filename] ... [filename]]"))
    print(_("Option(s):"))
    print(_("            -d level  - Debug level 1-10 (0 silent; 1 some; 10 lots)"))
    print(_("            -j pname  - Load project (saved session)"))
    print(_("            -v        - Verbose (to stdout and log)"))
    print(_("            -f        - Start Full screen"))
    print(_("            -c        - Dump Config"))
    print(_("            -o        - Use real stdout (for debug strings)"))
    print(_("            -V        - Show version"))
    print(_("            -x        - Clear (eXtinguish) config (will prompt)"))
    print(_("            -k        - Show Keys Presses"))
    print(_("            -h        - Help"))
    print()
    sys.exit(1)

# ------------------------------------------------------------------------

#def terminate(arg1 = None, arg2 = None):
def terminate():
    ''' Termination Handler'''
    if pedconfig.conf.verbose:
        print(_("Terminating pydepro.py, saving files to ~/pydepro"))

    # Save all
    pedconfig.conf.pedwin.activate_quit(None)
    #return signal.SIG_IGN

# ------------------------------------------------------------------------
# Start of program:


if __name__ == '__main__':

    ''' Main Entry Point for the editor '''

    # Redirect stdout to a fork to real stdout and log. This way messages can
    # be seen even if pydepro is started without a terminal (from the GUI)

    opts = []; args = []

    try:
        opts, args = getopt.getopt(sys.argv[1:], "d:h?vVj:fxctok",
                        ["debug=", "help", "help", "verbose", "version", "project="])

    except getopt.GetoptError as err:
        print(_("Invalid option(s) on command line:"), err)
        sys.exit(1)

    #print ("opts", opts, "args", args)
    #sys.exit(0)

    pedconfig.conf.version = VERSION
    pedconfig.conf.build_date = BUILDDATE
    pedconfig.conf.progname = PROGNAME

    pname = ""
    # Outdated parsing ... for now, leave it as is
    for aa in opts:
        #print("opt", aa[0])
        if aa[0] == "-d" or aa[0] == "--debug":
            try:
                pedconfig.conf.pgdebug = int(aa[1])
                print( PROGNAME, _("Running at debug level:"),  pedconfig.conf.pgdebug)
            except:
                pedconfig.conf.pgdebug = 0

        if aa[0] == "-j" or aa[0] == "--project":
            rrr = False
            try:
                if pedconfig.conf.verbose:
                    print( PROGNAME, "Loading sess/project", aa[1])
                pname = pedconfig.conf.sess_data + os.sep + aa[1]
                if not os.path.isfile(pname):
                    pname = pedconfig.conf.sess_data + os.sep + aa[1] + ".sess"
                    if not os.path.isfile(pname):
                        rrr = True
            except:
                print("Exception on loading project", sys.exc_info())
                #pedutil.put_exception("load proj")

            if rrr:
                print( PROGNAME, "Error on loading project", pname)
                sys.exit(1)

        if aa[0] == "-h" or  aa[0] == "--help" or aa[0] == "-?":
            xhelp()
        if aa[0] == "-V" or aa[0] == "--version":
            xversion()
        if aa[0] == "-v" or aa[0] == "--verbose":
            pedconfig.conf.verbose = True
        if aa[0] == "-f":
            pedconfig.conf.full_screen = True
        if aa[0] == "-v":
            pedconfig.conf.verbose = True
        if aa[0] == "-x":
            CLEAR_CONFIG = True
        if aa[0] == "-c":
            SHOW_CONFIG = True
        if aa[0] == "-t":
            SHOW_TIMING = True
        if aa[0] == "-o":
            USE_STDOUT = True
        if aa[0] == "-k":
            pedconfig.conf.show_keys = True
            #print("Showing keys")

    if pedconfig.conf.pgdebug > 0:
        print("Running '{}'".format(os.path.abspath(sys.argv[0])) )

    try:
        if not os.path.isdir(pedconfig.conf.config_dir):
            if pedconfig.conf.verbose:
                print("making", pedconfig.conf.config_dir)
            os.mkdir(pedconfig.conf.config_dir)
    except:
        pass

    # Let the user know if it needs fixin'
    if not os.path.isdir(pedconfig.conf.config_dir):
        print(_("Cannot access config dir:"), pedconfig.conf.config_dir)
        sys.exit(1)

    pedconfig.ensure_dirs(pedconfig.conf)

    if pedconfig.conf.verbose:
        print(_("Data stored in "), pedconfig.conf.config_dir)

    # Initialize sqlite to load / save preferences & other info
    # Initialize pedconfig for use

    pedconfig.conf.sql = pedsql.pedsql(pedconfig.conf.sql_data)
    pedconfig.conf.mydir = os.path.abspath(__file__)
    #print("Exe path:",  pedconfig.conf.mydir)

    # To clear all config vars
    if CLEAR_CONFIG:
        print(_("Are you sure you want to clear config ? (y/n)"))
        sys.stdout.flush()
        aa = sys.stdin.readline()
        if aa[0] == "y":
            print(_("Removing configuration ... "), end=' ')
            pedconfig.conf.sql.rmall()
            print("OK")
        sys.exit(0)

    # To check all config vars
    if SHOW_CONFIG:
        print("Dumping configuration:")
        ss = pedconfig.conf.sql.getall()
        for aa in ss:
            print(aa)
        sys.exit(0)

    # Uncomment this for silent stdout
    if USE_STDOUT or pedconfig.conf.pgdebug or \
                    pedconfig.conf.verbose:
        # Do not hide console
        #print("Using real stdout")
        pedwin.hidden = True    # Already hidden no hide
    else:
        pedwin.hidden = False   # Take action, hide

    sys.stdout = pedlog.fake_stdout(sys.stdout)
    sys.stderr = pedlog.fake_stdout(sys.stdout)

    # Uncomment this for buffered output
    if pedconfig.conf.verbose:
        print("Started", PROGNAME, "in", pedconfig.conf.orig_dir)

    main(pname, args[0:])

# EOF
